1. Подберите скорость обучения (alpha) и количество итераций.

*3. вместо того, чтобы задавать количество итераций, задайте условие остановки алгоритма - когда ошибка за итерацию
начинает изменяться ниже определенного порога (упрощенный аналог параметра tol в линейной регрессии в sklearn).

In [1]:
import numpy as np

In [2]:
def calc_mse(y, y_pred):
  err = np.mean((y - y_pred)**2)
  return err

In [3]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])

In [4]:
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

In [5]:
lpha = 0.1
n = X.shape[1]
for alfa_iter in range(1, 9):
    alpha = 1 * 0.1**alfa_iter
    W = np.array([1, 0.5])
    err = np.inf
    err_pred = np.inf
    iter_count = 0
    while err <= err_pred and iter_count < 5000:
        iter_count += 1
        err_pred = err
        y_pred = np.dot(W, X)
        err = calc_mse(y, y_pred)
        for ii in range(W.shape[0]):
            W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    print(alpha, iter_count, W, round(err, 4))

0.1 2 [  3.156  -36.5104] 9346.1224
0.010000000000000002 1650 [47.23214143  3.91071455] 45.9375
0.0010000000000000002 5000 [46.98926025  3.95556442] 45.9693
0.00010000000000000002 5000 [21.01242611  8.75238226] 415.6122
1.0000000000000003e-05 5000 [4.89544393 9.08446051] 1090.2178
1.0000000000000004e-06 5000 [1.52732513 2.07512148] 2620.4124
1.0000000000000004e-07 5000 [1.05485271 0.66869421] 3110.2282
1.0000000000000005e-08 5000 [1.00550752 0.51698688] 3166.7735


In [6]:
alpha = 0.1
n = X.shape[1]
arr_result = []
for alfa_iter in range(1, 9):
    alpha = round(1 * 0.1/alfa_iter, 4)
    W = np.array([1, 0.5])
    err = np.inf
    err_pred = np.inf
    iter_count = 0
    while err <= err_pred and iter_count < 5000:
        iter_count += 1
        err_pred = err
        y_pred = np.dot(W, X)
        err = calc_mse(y, y_pred)
        for ii in range(W.shape[0]):
            W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    result = {'alpha': alpha, 'iter_count': iter_count, 'W': W, 'err': err}
    arr_result.append(result)
    print(alpha, iter_count, W, round(err, 4))

0.1 2 [  3.156  -36.5104] 9346.1224
0.05 346 [47.23214244  3.91071436] 45.9375
0.0333 516 [47.2321423   3.91071439] 45.9375
0.025 664 [47.23214174  3.91071449] 45.9375
0.02 850 [47.23214209  3.91071443] 45.9375
0.0167 1015 [47.23214202  3.91071444] 45.9375
0.0143 1166 [47.23214171  3.9107145 ] 45.9375
0.0125 1334 [47.2321417  3.9107145] 45.9375


При alpha = 0.05 и количестве итераций 346 быстрее всего достигается нужный коэфициент ошибки.